// vitepress-plugin-sidebar-resolve/dist/index.cjs
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

const node_fs = require('node:fs');
const node_path = require('node:path');
const matter = require('gray-matter');
const vite = require('vite');
const picocolors = require('picocolors');

function _interopDefaultCompat (e) { return e && typeof e === 'object' && 'default' in e ? e.default : e; }

const matter__default = /*#__PURE__*/_interopDefaultCompat(matter);
const picocolors__default = /*#__PURE__*/_interopDefaultCompat(picocolors);

const getTitleFromMarkdown = (mdContent) => {
  const lines = mdContent.trimStart().split(/\r?\n/);
  for (const line of lines) {
    if (line.startsWith("# ")) {
      return line.substring(2).trim();
    }
  }
  return void 0;
};
const isIllegalIndex = (index) => {
  return isNaN(index) || index < 0;
};
const isSome = (arr, name) => {
  return arr.some((item) => item === name || item instanceof RegExp && item.test(name));
};

const version = "1.1.2";

const logger = vite.createLogger("info", {
  prefix: `[vitepress-plugin-sidebar-resolve v${version}]`
});
const info = (message, level = "green", option = { timestamp: true }) => {
  logger.info(picocolors__default[level](message), option);
};
const warn = (message, level = "yellow", option = { timestamp: true }) => {
  logger.warn(picocolors__default[level](message), option);
};
const warnOnce = (message, level = "yellow", option = { timestamp: true }) => {
  logger.info(picocolors__default[level](message), option);
};
const error = (message, level = "red", option = { timestamp: true }) => {
  logger.error(picocolors__default[level](message), option);
};
const logger$1 = {
  info,
  warn,
  warnOnce,
  error
};

const DEFAULT_IGNORE_DIR = ["node_modules", "dist", ".vitepress", "public"];
const createSidebar = (option = {}, prefix = "/") => {
  const {
    path,
    ignoreList = [],
    scannerRootMd = true,
    collapsed,
    titleFormMd = false,
    initItems = true,
    initItemsText = false,
    sidebarResolved,
    ignoreWarn = false,
    indexSeparator,
    prefixTransform,
    suffixTransform
  } = option;
  if (!path) return {};
  prefix = prefix.replace(/\/$/, "") + "/";
  const sidebar = {};
  const dirPaths = readDirPaths(path, ignoreList);
  const key = prefix === "/" ? prefix : `/${prefix}`;
  if (scannerRootMd) sidebar[key] = createSidebarItems(path, { ...option, ignoreIndexMd: true }, key, scannerRootMd);
  dirPaths.forEach((dirPath) => {
    const fileName = node_path.basename(dirPath);
    const sidebarItems = createSidebarItems(dirPath, option, `${key}${fileName}/`);
    if (!ignoreWarn && !sidebarItems.length) {
      return logger$1.warn(`\u8BE5\u76EE\u5F55 '${dirPath}' \u5185\u90E8\u6CA1\u6709\u4EFB\u4F55\u6587\u4EF6\u6216\u6587\u4EF6\u5E8F\u53F7\u51FA\u9519\uFF0C\u5C06\u5FFD\u7565\u751F\u6210\u5BF9\u5E94\u4FA7\u8FB9\u680F`);
    }
    const { name, title } = resolveFileName(fileName, dirPath, indexSeparator);
    const info = getInfoFromMarkdown(dirPath, fileName);
    const mdTitle = titleFormMd ? info.title : "";
    const sidebarPrefix = (info.prefix && (prefixTransform?.(info.prefix) ?? info.prefix)) ?? "";
    const sidebarSuffix = (info.suffix && (suffixTransform?.(info.suffix) ?? info.suffix)) ?? "";
    const text = sidebarPrefix + (initItemsText ? mdTitle || title : "") + sidebarSuffix;
    sidebar[`${key}${fileName}/`] = initItems ? [
      {
        text,
        collapsed: typeof collapsed === "function" ? collapsed(prefix + name, text) : collapsed,
        items: sidebarItems
      }
    ] : sidebarItems;
  });
  return sidebarResolved?.(sidebar) ?? sidebar;
};
const readDirPaths = (sourceDir, ignoreList = []) => {
  const dirPaths = [];
  const ignoreListAll = [...DEFAULT_IGNORE_DIR, ...ignoreList];
  const dirOrFilenames = node_fs.readdirSync(sourceDir);
  dirOrFilenames.forEach((dirOrFilename) => {
    const secondDirPath = node_path.resolve(sourceDir, dirOrFilename);
    if (!isSome(ignoreListAll, dirOrFilename) && node_fs.statSync(secondDirPath).isDirectory()) {
      dirPaths.push(secondDirPath);
    }
  });
  return dirPaths;
};
const createSidebarItems = (root, option, prefix = "/", onlyScannerRootMd = false) => {
  const {
    collapsed,
    ignoreList = [],
    ignoreIndexMd = false,
    fileIndexPrefix = false,
    sidebarItemsResolved,
    beforeCreateSidebarItems,
    titleFormMd = false,
    ignoreWarn = false,
    sort = true,
    defaultSortNum = 9999,
    sortNumFromFileName = false,
    indexSeparator,
    prefixTransform,
    suffixTransform
  } = option;
  const ignoreListAll = [...DEFAULT_IGNORE_DIR, ...ignoreList];
  let sidebarItems = [];
  const sidebarItemsNoIndex = [];
  let dirOrFilenames = node_fs.readdirSync(root);
  dirOrFilenames = beforeCreateSidebarItems?.(dirOrFilenames) ?? dirOrFilenames;
  dirOrFilenames.forEach((dirOrFilename) => {
    if (isSome(ignoreListAll, dirOrFilename)) return [];
    const filePath = node_path.resolve(root, dirOrFilename);
    const { index: indexStr, title, type, name } = resolveFileName(dirOrFilename, filePath, indexSeparator);
    const index = parseInt(indexStr, 10);
    if (!ignoreWarn && fileIndexPrefix && isIllegalIndex(index)) {
      logger$1.warn(`\u8BE5\u6587\u4EF6 '${filePath}' \u5E8F\u53F7\u51FA\u9519\uFF0C\u8BF7\u586B\u5199\u6B63\u786E\u7684\u5E8F\u53F7`);
      return [];
    }
    if (!ignoreWarn && sidebarItems[index]) logger$1.warn(`\u8BE5\u6587\u4EF6 '${filePath}' \u7684\u5E8F\u53F7\u5728\u540C\u4E00\u7EA7\u522B\u4E2D\u91CD\u590D\u51FA\u73B0\uFF0C\u5C06\u4F1A\u88AB\u8986\u76D6`);
    if (!onlyScannerRootMd && node_fs.statSync(filePath).isDirectory()) {
      const info = getInfoFromMarkdown(root, dirOrFilename);
      const mdTitle = titleFormMd ? info.title : "";
      const sidebarPrefix = (info.prefix && (prefixTransform?.(info.prefix) ?? info.prefix)) ?? "";
      const sidebarSuffix = (info.suffix && (suffixTransform?.(info.suffix) ?? info.suffix)) ?? "";
      const text = sidebarPrefix + (mdTitle || title) + sidebarSuffix;
      const childSidebarItems = createSidebarItems(filePath, option, `${prefix}${dirOrFilename}/`);
      let sidebarItem = {
        text,
        collapsed: typeof collapsed === "function" ? collapsed(prefix + name, text) : collapsed,
        items: childSidebarItems
      };
      if (sort) {
        sidebarItem = {
          ...sidebarItem,
          // 对子侧边栏进行排序
          items: childSidebarItems.sort((a, b) => (a.sort || defaultSortNum) - (b.sort || defaultSortNum)).map((item) => {
            delete item.sort;
            return item;
          }),
          sort: sortNumFromFileName ? index : info.sort
        };
      }
      if (isIllegalIndex(index)) sidebarItemsNoIndex.push(sidebarItem);
      else sidebarItems[index] = sidebarItem;
    } else {
      if (onlyScannerRootMd && dirOrFilename === "index.md") return [];
      if (ignoreIndexMd && ["index.md", "index.MD"].includes(dirOrFilename)) return [];
      if (!["md", "MD"].includes(type)) {
        if (!ignoreWarn && !onlyScannerRootMd) logger$1.warn(`\u8BE5\u6587\u4EF6 '${filePath}' \u975E .md \u683C\u5F0F\u6587\u4EF6\uFF0C\u4E0D\u652F\u6301\u8BE5\u6587\u4EF6\u7C7B\u578B`);
        return [];
      }
      const content = node_fs.readFileSync(filePath, "utf-8");
      const {
        data: { title: frontmatterTitle, sidebar = true, sidebarSort, sidebarPrefix, sidebarSuffix } = {},
        content: mdContent
      } = matter__default(content, {});
      if (!sidebar) return [];
      const mdTitle = titleFormMd ? getTitleFromMarkdown(mdContent) : "";
      const finalSidebarPrefix = (sidebarPrefix && (prefixTransform?.(sidebarPrefix) ?? sidebarPrefix)) ?? "";
      const finalSidebarSuffix = (sidebarSuffix && (suffixTransform?.(sidebarSuffix) ?? sidebarSuffix)) ?? "";
      const text = finalSidebarPrefix + (frontmatterTitle || mdTitle || title) + finalSidebarSuffix;
      let sidebarItem = {
        text,
        collapsed: typeof collapsed === "function" ? collapsed(prefix + name, text) : collapsed,
        link: prefix + name
      };
      if (sort) sidebarItem = { ...sidebarItem, sort: sortNumFromFileName ? index : sidebarSort };
      if (isIllegalIndex(index)) sidebarItemsNoIndex.push(sidebarItem);
      else sidebarItems[index] = sidebarItem;
    }
  });
  sidebarItems = [...sidebarItems, ...sidebarItemsNoIndex].filter(Boolean);
  if (sort) {
    sidebarItems = sidebarItems.sort((a, b) => (a.sort || defaultSortNum) - (b.sort || defaultSortNum)).map((item) => {
      delete item.sort;
      return item;
    });
  }
  return sidebarItemsResolved?.(sidebarItems) ?? sidebarItems;
};
const getInfoFromMarkdown = (root, dirOrFilename) => {
  const state = {
    title: void 0,
    sort: void 0,
    prefix: "",
    suffix: ""
  };
  const filePaths = [
    node_path.join(root, dirOrFilename, "index.md"),
    node_path.join(root, dirOrFilename, "index.MD"),
    node_path.join(root, dirOrFilename, dirOrFilename + ".md")
  ];
  for (const filePath of filePaths) {
    if (!node_fs.existsSync(filePath)) continue;
    const content = node_fs.readFileSync(filePath, "utf-8");
    const { data: { title, sidebarSort, sidebarPrefix, sidebarSuffix } = {}, content: mdContent } = matter__default(content, {});
    const t = title || getTitleFromMarkdown(mdContent);
    if (!state.title) state.title = t;
    if (!state.sort) state.sort = sidebarSort;
    if (!state.prefix) state.prefix = sidebarPrefix;
    if (!state.suffix) state.suffix = sidebarSuffix;
  }
  return state;
};
const resolveFileName = (filename, filePath, separator = ".") => {
  const stat = node_fs.statSync(filePath);
  if (separator !== "." && isExtraSeparator(filename, separator)) {
    return parseExtraSeparator(filename, stat.isDirectory(), separator);
  }
  if (filename.includes(".")) {
    return parseDotSeparator(filename, stat.isDirectory());
  }
  return { index: "", title: filename, type: "", name: filename };
};
const parseDotSeparator = (filename, isDirectory) => {
  const parts = filename.split(".");
  if (parts.length === 2) {
    const index = parts[0] === "index" ? "0" : parts[0];
    const title = isDirectory ? parts[1] : parts[0];
    const type = isDirectory ? "" : parts[1];
    const name = parts[0];
    return { index, title, type, name };
  } else {
    const firstDotIndex = filename.indexOf(".");
    const lastDotIndex = filename.lastIndexOf(".");
    const index = filename.substring(0, firstDotIndex);
    const title = filename.substring(firstDotIndex + 1, lastDotIndex);
    const type = isDirectory ? "" : filename.substring(lastDotIndex + 1);
    const name = isDirectory ? filename : filename.substring(0, lastDotIndex);
    return { index, title, type, name };
  }
};
const isExtraSeparator = (filename, separator) => {
  if (!filename.includes(separator)) return false;
  const parts = filename.split(separator, 2);
  if (!/^\d+$/.test(parts[0])) return false;
  return true;
};
const parseExtraSeparator = (filename, isDirectory, separator) => {
  const firstSeparatorIndex = filename.indexOf(separator);
  const lastDotIndex = filename.lastIndexOf(".");
  const index = filename.substring(0, firstSeparatorIndex);
  const title = isDirectory ? filename.substring(firstSeparatorIndex + 1) : filename.substring(firstSeparatorIndex + 1, lastDotIndex);
  const type = isDirectory ? "" : filename.substring(lastDotIndex + 1);
  const name = isDirectory ? filename : filename.substring(0, lastDotIndex);
  return { index, title, type, name };
};

function VitePluginVitePressSidebarResolve(option = {}) {
  let isExecute = false;
  return {
    name: "vite-plugin-vitepress-sidebar-resolve",
    configureServer({ watcher, restart }) {
      if (!option.restart) return;
      watcher.add("*.md");
      watcher.on("add", async (path) => {
        if (!path.endsWith(".md")) return;
        await restart();
      }).on("unlink", async (path) => {
        if (!path.endsWith(".md")) return;
        await restart();
      });
    },
    config(config) {
      if (isExecute) return;
      isExecute = true;
      const {
        site: { themeConfig = {}, locales = {} },
        srcDir
      } = config.vitepress;
      const { path, ignoreList, localeRootDir } = option;
      const baseDir = path ? node_path.join(srcDir, path) : srcDir;
      const localesKeys = Object.keys(locales).filter((key) => key !== "root");
      if (!localesKeys.length) return setSideBar(themeConfig, createSidebar({ ...option, path: baseDir }));
      localesKeys.forEach((localesKey) => {
        const sidebar = createSidebar(
          { ...option, path: `${baseDir}/${localesKey}` },
          localesKey
        );
        setSideBar(locales[localesKey].themeConfig, sidebar);
      });
      const rootDir = localeRootDir ? `/${localeRootDir}` : "";
      const rootSideBar = createSidebar({
        ...option,
        path: `${baseDir}${rootDir}`,
        ignoreList: [...ignoreList || [], ...localesKeys]
      });
      setSideBar(locales["root"].themeConfig, rootSideBar);
    }
  };
}
const setSideBar = (themeConfig, sidebar) => {
  themeConfig = themeConfig || {};
  themeConfig.sidebar = {
    ...sidebar,
    ...Array.isArray(themeConfig.sidebar) ? logger$1.warn("\u81EA\u5B9A\u4E49 Sidebar \u5FC5\u987B\u662F\u5BF9\u8C61\u5F62\u5F0F") : themeConfig.sidebar
  };
  logger$1.info("Injected Sidebar Data Successfully. \u6CE8\u5165\u4FA7\u8FB9\u680F\u6570\u636E\u6210\u529F!");
};

exports.default = VitePluginVitePressSidebarResolve;
exports.getTitleFromMarkdown = getTitleFromMarkdown;
exports.isIllegalIndex = isIllegalIndex;
exports.isSome = isSome;
